### ISSUE TITLE
Provide hooks for instrumentation of server workings

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#35]
Acceptance: we should be able to integrate metrics without a core dependency on metrics.

(Edit: we've decided an http4s-server dependency on Metrics is acceptable.)

**Comment by dchenbecker** [comment]
Any estimate on level of effort for this, or ETA? We're very interested in http4s, but we need solid metrics (but not necessarily Coda's Metrics). We'd like to help if this doesn't involve a major architecture change.

**Comment by bryce-anderson** [comment]
Though we don't have an official ETA, its been coming up a lot recently. We'd very much appreciate help in the form of discussion and/or code.
I have a feeling that the depth of changes required will be correlated with the types of metrics needed, but a metrics middleware will be able to cover a lot of ground and will have almost no effect on existing infrastructure.
Feel free drop thoughts/needs here or in the #http4s irc channel.

**Comment by rossabaker** [comment]
A port of metrics-servlets would benefit all backends whether we provided a middleware or finally solved the issues raised in #85.  That seems like a first step that adds value regardless of how we solve the harder problem.

**Comment by bryce-anderson** [comment]
+1, its relatively easy and better than nothing.
One minor complicated arises: we need to serialize the data somehow.
Metrics provides serializers for the java Jackson library, but that could
end up conflicting with json4s-jackson. Its just something we will need to
keep an eye on.

**Comment by rossabaker** [comment]
Are you proposing a dependency on metrics-json?  That's the easiest road, with the downside you mention.  Alternatively, could reimplement MetricsModule on top of a JSON utility class.

**Comment by bryce-anderson** [comment]
Using metrics-json would be my first inclination. If problems arise just
continue down the road to our own JSON class.

**Comment by rossabaker** [comment]
I'm less worried about compatibility problems than classpath bloat.  So, +1, let's start like that, and the first person to complain of classpath bloat is cordially invited to send a pull request. :)

**Comment by rossabaker** [comment]
Though https://github.com/json4s/json4s/issues/186 makes me worry about Jackson compatibility.  Ugh.

**Comment by rossabaker** [comment]
We have pull requests in for the last tickets tagged for 0.6, so let's prioritize this in 0.7.

@dchenbecker: "Not necessarily Coda's metrics" means you'd be happy with any kind of metrics, or you don't want to tightly couple http4s' metrics to Coda?  I'd like to abstract it, but I also love Coda's and don't want to reinvent the flat tire.

**Comment by dchenbecker** [comment]
What I meant by "not necessarily Coda's metrics" was that I would be happy with any similar metrics implementation, not that I don't want you to use Coda's stuff. I *do* think that using Coda's lib as a start will be an immediate win, and can always be revisited down the road if it becomes problematic. We use Argonaut for JSON, so I don't think that jackson compatibility is an issue for us specifically. Similarly, we're using Play for simple REST services, which is kinda like using a shipping container to carry groceries, so bloat is the least of my concerns.

In any case, I'd be happy to hash out more details on IRC if you'd like, but it sounds like metrics-servlet would be a pretty easy thing to integrate.

**Comment by rossabaker** [comment]
Great, I thought that's what you meant, and I agree.

So this problem breaks down into two parts:

1. Collecting the metrics.
    1. A middleware, which I think @bryce-anderson may have laying around, that is portable, but does not measure the overhead of the backend.
    2. metrics-servlet, which wraps Http4sServlet.  A little better measurement than #1, but we can do better, and Blaze would need to roll its own.
    3. metrics-jetty9.  This offers a more thorough measurement of request timings and a richer set of metrics.  Installing this impacts how a Jetty server is constructed, which means an http4s-jetty-metrics module or an optional dependency in http4s-jetty.
    4. An abstraction of a Jetty Handler, a Tomcat valve, and a Blaze stage that would give us a more portable solution that gets to the outermost edges.  This would be neat, but is also the one we've spun our wheels on.
2. Exposing the metrics.
   1. metrics-servlets would just work on the servlet implementations.
   2. Porting those servlets to http4s services would be straightforward and work for blaze.
   3. Could make an eye candy dashboard, but first things first: I want the JSON.

We need at least one from section 1 and one from section 2 to have metrics.  Let's pick the work we want to undertake for 0.7, create subtasks with owners so we're not duplicating work, and get something working.

**Comment by rossabaker** [comment]
Just to see what could be done with minimal effort, I took the metrics-jetty9 and metrics-servlets route.  Check out the topic/metrics branch, 'sbt examples-jetty/run', and enjoy the http://localhost:8080/metrics.  It's incomplete and icky, but at least now it's tangible.

**Comment by rossabaker** [comment]
Going down this road is really easy, but we don't get equivalent metrics names across the backends.  To do that, we'd need to define our own contract, and then implement it atop the servlet backends individually, as we're going to have to do for Blaze anyway.

**Comment by rossabaker** [comment]
This is the JSON we get now, for comparison's sake:

### Tomcat

```json
{
    "counters": {
        "com.codahale.metrics.servlet.InstrumentedFilter.activeRequests": {
            "count": 1
        }
    }, 
    "gauges": {}, 
    "histograms": {}, 
    "meters": {
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.badRequest": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.created": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.noContent": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.notFound": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.ok": {
            "count": 16, 
            "m15_rate": 0.060798242160400036, 
            "m1_rate": 2.731150416995205e-09, 
            "m5_rate": 0.005562553986732153, 
            "mean_rate": 0.013885127224030335, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.other": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.serverError": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }
    }, 
    "timers": {
        "com.codahale.metrics.servlet.InstrumentedFilter.requests": {
            "count": 16, 
            "duration_units": "seconds", 
            "m15_rate": 0.060798242160400036, 
            "m1_rate": 2.731150416995205e-09, 
            "m5_rate": 0.005562553986732153, 
            "max": 0.20785610100000002, 
            "mean": 0.016175323411380747, 
            "mean_rate": 0.013885128242081641, 
            "min": 0.0006300040000000001, 
            "p50": 0.0007660970000000001, 
            "p75": 0.000983883, 
            "p95": 0.20785610100000002, 
            "p98": 0.20785610100000002, 
            "p99": 0.20785610100000002, 
            "p999": 0.20785610100000002, 
            "rate_units": "calls/second", 
            "stddev": 0.05082027904991503
        }
    }, 
    "version": "3.0.0"
}
```

### Jetty

```json
{
    "counters": {
        "org.eclipse.jetty.servlet.ServletContextHandler.active-dispatches": {
            "count": 1
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.active-requests": {
            "count": 1
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.active-suspended": {
            "count": 0
        }
    }, 
    "gauges": {}, 
    "histograms": {}, 
    "meters": {
        "org.eclipse.jetty.servlet.ServletContextHandler.1xx-responses": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.2xx-responses": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.3xx-responses": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.4xx-responses": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.5xx-responses": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.async-dispatches": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.async-timeouts": {
            "count": 0, 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "mean_rate": 0.0, 
            "units": "events/second"
        }
    }, 
    "timers": {
        "http.connections": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.connect-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.delete-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.dispatches": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.get-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.head-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.move-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.options-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.other-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.post-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.put-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }, 
        "org.eclipse.jetty.servlet.ServletContextHandler.trace-requests": {
            "count": 0, 
            "duration_units": "seconds", 
            "m15_rate": 0.0, 
            "m1_rate": 0.0, 
            "m5_rate": 0.0, 
            "max": 0.0, 
            "mean": 0.0, 
            "mean_rate": 0.0, 
            "min": 0.0, 
            "p50": 0.0, 
            "p75": 0.0, 
            "p95": 0.0, 
            "p98": 0.0, 
            "p99": 0.0, 
            "p999": 0.0, 
            "rate_units": "calls/second", 
            "stddev": 0.0
        }
    }, 
    "version": "3.0.0"
}
```

**Comment by dchenbecker** [comment]
It looks like they're comparable for the basics. I'm not so worried about the different names, although I agree it would be nice to have things unified. Do you think the short-term work for metrics-(jetty9|servlet) is feasible, or is it really important to abstract to a common interface?

**Comment by rossabaker** [comment]
I think I'd rather deliver incremental value now and build our architectural wonder later.

It turns out there's very little overlap between servlet's InstrumentedFilter and jetty9's InstrumentedHandler.  That surprised me.

**Comment by dchenbecker** [comment]
OK, awesome. I thought that's where you were headed but then your last comment worried me a bit.

**Comment by bryce-anderson** [comment]
I see nothing in the Tomcat JSON and only a few in the Jetty JSON that can't be done with a middleware, the most notable being timeouts.

**Comment by rossabaker** [comment]
The problem with timers in a middleware is they don't measure the time spent in the backend.  By instrumenting the Jetty handler (best) or using a servlet filter (good), the time spent parsing the request and rendering the response*  is included.  Doing it this way trades portability for accuracy.  It would be interesting to compare a middleware's numbers to backend-native numbers.

* - I suppose the response rendering time might be absorbed in a middleware with an `eval_(timerContext.stop())` in an `onFinish` handler tacked onto the response body.

**Comment by rossabaker** [comment]
A crude comparison of metrics-jetty9 vs. middleware.  10000 pings, concurrency 100.

InstrumentedHandler, patched from #204:

```json
        "org.http4s.server.get-requests": {
            "count": 10000, 
            "duration_units": "seconds", 
            "m15_rate": 8.49970119088744, 
            "m1_rate": 122.6678590756922, 
            "m5_rate": 25.358095546874445, 
            "max": 0.528, 
            "mean": 0.009522693532747794, 
            "mean_rate": 529.8538648967696, 
            "min": 0.0, 
            "p50": 0.005, 
            "p75": 0.01, 
            "p95": 0.023000000000000003, 
            "p98": 0.030000000000000002, 
            "p99": 0.033, 
            "p999": 0.523, 
            "rate_units": "calls/second", 
            "stddev": 0.03231633203613846
        }, 
```

Middleware, as seen on 5f3886556b76c34cb968bb400f4888ef132d622a:

```json
        "middleware.get-requests": {
            "count": 10000, 
            "duration_units": "seconds", 
            "m15_rate": 6.694719670882792, 
            "m1_rate": 96.61832936192573, 
            "m5_rate": 19.97309520195742, 
            "max": 0.15399579500000002, 
            "mean": 0.00036656059233993224, 
            "mean_rate": 520.0337939155863, 
            "min": 4.6109e-05, 
            "p50": 8.4342e-05, 
            "p75": 0.000104016, 
            "p95": 0.00023528600000000002, 
            "p98": 0.001145317, 
            "p99": 0.003344066, 
            "p999": 0.01941413, 
            "rate_units": "calls/second", 
            "stddev": 0.004863065922628225
        }
```

**Comment by rossabaker** [comment]
Six years of inaction suggests to me that we've all learned to live with the basic accuracy vs. portability tradeoff and are picking our poison.
